### DISCUSSION TITLE
Array of different schemas

### DISCUSSION THREAD

**Discussion by vnugent** [Automattic/mongoose#12154] [Q&A]
My `audit_logs` collection tracks changes to a few collections (via change streams).  Depending on the user actions, the `changes` array can contain documents with the same schema or different schemas.  In this example, profile and account have a different schema.

How do you define the schema for `changes`?  Using array of Mixed type? (https://mongoosejs.com/docs/schematypes.html#mixed)

```javascript
# audit_logs collection
[
    {
      _id: new ObjectId("62dd9815b11c3e3e3d01a41c"),
      editedBy: new Binary(Buffer.from("a927c2893cc64963b67f560c81a5dffa", "hex"), 4),
      operation: 'addNewUser',
      createdAt: 2022-07-24T19:05:57.855Z,
      changes: [
           {
               # 'profile' fullDocument captured by  event stream
           },
           {
               # 'account' fullDocument captured by event stream 
           }
      ]
    }
   ...
]

```

**Comment by vkarpov15** [comment]
Do you know the model's type in `audit_logs`? If you do, you can use discriminators, or manually `hydrate()` the `changes` subdocuments. If not, need to use mixed.

**Reply by vnugent** [reply]
Yes, I have a schema defined for `audit_log`.  The reason I'm wanting to embed `changes` is to avoid join/lookup multiple collections.  

I've just looked at [Discriminators docs](https://mongoosejs.com/docs/discriminators.html).  I'm not sure how to use it as an array.

**Comment by vkarpov15** [comment]
For this use case, I think you're better off storing `changes` as an array of mixed. `new Schema({ changes: [] })`. And then hydrate the individual documents using `const account = Account.hydrate(auditLog.changes[0]); // assuming changes[0] is an account` . Discriminators would be a bit trickier to implement because you have a bunch of top-level models with potentially different options.
